### PR TITLE
HOSTEDCP-1513: Add operator scoping support to nodepool controller

### PR DESCRIPTION
**Pull request by lucaspalm**
**What this PR does / why we need it**:
These changes extend the hostedcluster annotation scoping support to the nodepool controller.  This feature was originally introduced in https://github.com/openshift/hypershift/pull/3702, but support was only added to the hostedcluster controller at that time.

**Which issue(s) this PR fixes**:
https://issues.redhat.com//browse/HOSTEDCP-1513

### CODE REVIEW

File: support/util/util.go [hunk 1]
```diff
@@ -458,3 +458,109 @@ func PredicatesForHostedClusterChildResourcesAnnotationScoping(r client.Reader)
...
 	return predicate.NewPredicateFuncs(filter)
 }
+
+// PredicatesForNodepoolAnnotationScoping returns predicate filters for all event types that will ignore incoming
+// event requests for nodepool resources in which their owning hostedcluster resource doesn't have a scope annotation
+// that matches what is specified in the HOSTEDCLUSTERS_SCOPE_ANNOTATION env var.  If not defined or empty, the default behavior
+// is to accept all nodepool events in which the owning hostedcluster resource does not have a corresponding scope annotation defined.
+// The ENABLE_HOSTEDCLUSTERS_ANNOTATION_SCOPING env var must also be set to "true" to enable the scoping feature.
+func PredicatesForNodepoolAnnotationScoping(r client.Reader) predicate.Predicate {
+	hcAnnotationScopingEnabledEnvVal := os.Getenv(EnableHostedClustersAnnotationScopingEnv)
```
Comment (stevekuznetsov) [inline support/util/util.go:468]: nit: let's prefer to read these values at startup or as close to startup in `main()` as we can and pass them via normal function parameters to aid in understanding control flow

File: support/util/util.go [hunk 2]
```diff
@@ -458,3 +458,109 @@ func PredicatesForHostedClusterChildResourcesAnnotationScoping(r client.Reader)
...
+
+		// use the Cluster Name from the nodepool spec to get the owning hostedcluster object
+		hostedClusterName := ""
+		if np.Spec.ClusterName != "" {
+			hostedClusterName = np.Spec.ClusterName
+		}
+		if hostedClusterName == "" {
+			return true
+		}
+		namespacedName := ParseNamespacedName(fmt.Sprintf("%s/%s", np.Namespace, hostedClusterName))
```
Comment (stevekuznetsov) [inline support/util/util.go:488]: You can just create a `types.NamespacedName{}` literal using the fields - no need to string-format and parse.

File: support/util/util.go [hunk 3]
```diff
@@ -458,3 +458,109 @@ func PredicatesForHostedClusterChildResourcesAnnotationScoping(r client.Reader)
 	}
 	return predicate.NewPredicateFuncs(filter)
 }
+
+// PredicatesForNodepoolAnnotationScoping returns predicate filters for all event types that will ignore incoming
+// event requests for nodepool resources in which their owning hostedcluster resource doesn't have a scope annotation
+// that matches what is specified in the HOSTEDCLUSTERS_SCOPE_ANNOTATION env var.  If not defined or empty, the default behavior
+// is to accept all nodepool events in which the owning hostedcluster resource does not have a corresponding scope annotation defined.
+// The ENABLE_HOSTEDCLUSTERS_ANNOTATION_SCOPING env var must also be set to "true" to enable the scoping feature.
+func PredicatesForNodepoolAnnotationScoping(r client.Reader) predicate.Predicate {
```
Comment (stevekuznetsov) [inline support/util/util.go:467]: If we are using the state of the `HostedCluster` to determine whether or not to process a `NodePool`, we need to re-queue affected `NodePool`s when the `HostedCluster` changes (e.g. it gains or loses the annotation we're checking for)

File: support/util/util.go [hunk 4]
```diff
@@ -458,3 +458,109 @@ func PredicatesForHostedClusterChildResourcesAnnotationScoping(r client.Reader)
...
+	}
+	return predicate.NewPredicateFuncs(filter)
+}
+
+// PredicatesForNodepoolChildResourcesAnnotationScoping returns predicate filters for all event types that will ignore incoming
+// event requests for resources in which the parent hostedcluster does not
+// match the "scope" annotation specified in the HOSTEDCLUSTERS_SCOPE_ANNOTATION env var.  If not defined or empty, the
+// default behavior is to accept all events for hostedclusters that do not have the annotation.
+// The ENABLE_HOSTEDCLUSTERS_ANNOTATION_SCOPING env var must also be set to "true" to enable the scoping feature.
+func PredicatesForNodepoolChildResourcesAnnotationScoping(r client.Reader) predicate.Predicate {
```
Comment (stevekuznetsov) [inline support/util/util.go:515]: Could you DRY out these helpers by passing in the function to extract the `HostedCluster`'s `NamespacedName` from the `obj client.Object`?

File: support/util/util.go (hunk 3 above)
Comment (lucaspalm) [reply support/util/util.go:467]: I believe this is already happening, right?  In the hypothetical 2 operator topology, where "operator1" manages hostedclusters that do not have the specified annotation, and "operator2" is configured to manage hosteclusters that do have the annotation, as scope annotations are added/removed from hosteclusters, one of the operators should in-turn queue the NodePools.

File: support/util/util.go (hunk 4 above)
Comment (lucaspalm) [reply support/util/util.go:515]: What function are you referring to exactly?  Is there one that exists already somewhere or are you saying to create one use that?

Comment (stevekuznetsov) [reply support/util/util.go:515]: The latter - it seemed like the only difference between these utilities

Comment (stevekuznetsov) [reply support/util/util.go]: Nice! This looks really good!

**Comment by csrwng** [comment]
/override "Red Hat Konflux / hypershift-main-enterprise-contract / hypershift-operator-main"

**Pull request merged**